react/OrderShipping.tsx

```diff
@@ -1,31 +1,206 @@
-import React, { createContext, ReactNode, useContext } from 'react'
-
+import React, {
+  createContext,
+  ReactNode,
+  useContext,
+  useCallback,
+  useEffect,
+  useRef,
+} from 'react'
+import { compose, graphql } from 'react-apollo'
+import { useOrderQueue } from 'vtex.order-manager/OrderQueue'
 import { useOrderForm } from 'vtex.order-manager/OrderForm'
 
+import {
+  estimateShipping as EstimateShipping,
+  selectDeliveryOption as SelectDeliveryOption,
+} from 'vtex.checkout-resources/Mutations'
+
 interface Context {
   countries: string[]
+  selectedAddress: CheckoutAddress
+  insertAddress: any
+  deliveryOptions: DeliveryOption[]
+  selectDeliveryOption: any
 }
 
 interface OrderShippingProps {
   children: ReactNode
+  EstimateShipping: any
+  SelectDeliveryOption: any
 }
 
 const OrderShippingContext = createContext<Context | undefined>(undefined)
 
-export const OrderShippingProvider = ({ children }: OrderShippingProps) => {
-  const {
-    orderForm: {
-      shipping: { countries },
-    },
-  } = useOrderForm()
-
-  return (
-    <OrderShippingContext.Provider value={{ countries }}>
-      {children}
-    </OrderShippingContext.Provider>
-  )
+const shippingId = 'Shipping'
+
+const PENDING = 'Pending'
+const FULFILLED = 'Fulfilled'
+const TASK_CANCELLED = 'TASK_CANCELLED'
+
+const changeSelectedDeliveryOption = (
+  deliveryOptions: DeliveryOption[],
+  selectedDeliveryOptionId: string
+) => {
+  return deliveryOptions.map(option => {
+    if (option.isSelected) {
+      option.isSelected = false
+    } else if (option.id === selectedDeliveryOptionId) {
+      option.isSelected = true
+    }
+
+    return option
+  })
+}
+
+const updateShipping = (totalizers: Totalizer[], newShippingValue: number) => {
+  return totalizers.map(totalizer => {
+    if (totalizer.id === shippingId) {
+      totalizer.value = newShippingValue
+    }
+    return totalizer
+  })
 }
 
+const getShipping = (totalizers: Totalizer[]) => {
+  return totalizers.find(totalizer => {
+    return totalizer.id === shippingId
+  })
+}
+
+const findDeliveryOptionById = (
+  deliveryOptions: DeliveryOption[],
+  deliveryOptionId: string
+): DeliveryOption => {
+  return deliveryOptions.find(option => {
+    return option.id === deliveryOptionId
+  })!
+}
+
+export const OrderShippingProvider = compose(
+  graphql(EstimateShipping, { name: 'EstimateShipping' }),
+  graphql(SelectDeliveryOption, { name: 'SelectDeliveryOption' })
+)(
+  ({
+    children,
+    EstimateShipping,
+    SelectDeliveryOption,
+  }: OrderShippingProps) => {
+    const { enqueue, listen } = useOrderQueue()
+    const { orderForm, setOrderForm } = useOrderForm()
+
+    const {
+      shipping: { countries, selectedAddress, deliveryOptions },
+    } = orderForm
+
+    const isQueueBusy = useRef(false)
+    useEffect(() => {
+      const unlisten = listen(PENDING, () => (isQueueBusy.current = true))
+      return unlisten
+    })
```

Comment:
Without the dependency array, this will be "listening" and "unlistening" every time your component is updated. Is that what you want?

If your desire is to "unlisten" on unmounting the component, just add an empty array.
```suggestion
    }, [])
```

Reply:
I think in that case, the right thing to do is to put `listen` in the dependency array. Thanks!

Reply:
I feel like these listening `useEffect`s should moved to a custom hook in the `OrderQueue` repo. We have the exact same code in `OrderItems`...